### PR TITLE
Migrating to 0.3.+ and adding Broker instructions

### PR DESCRIPTION
**Pull request by danieldobalian**

### CODE REVIEW

**Comment by NLLAPPS** [comment]
Hi, what is the ETA for this release? I have checked https://aka.ms/MobileAppReg but it is redirecting to Docs at the moment.

I was just completing my implementation and would like to mive to version 0.3+ before I publish my app

**Comment by iambmelt** [comment]
Initial clone and build experience yields red errors that may disorient developers...

After a fresh clone + Open w/ Android Studio:
```
The value of attribute "android:host" associated with an element type "data" must not contain the '<' character.
```
Because the Gradle settings fail to sync, the project view displays a list of files, rather than a friendly project structure. The issue is the placeholders in the `AndroidManifest.xml`: they contain illegal characters.

The `host` and `scheme` attributes also seem to be flipped, depending on what we anticipate developers will register in the portal...

In the sample...
```xml
<!--Add in your scheme/host from registered redirect URI-->
<data
    android:scheme="msauth"
    android:host="<PACKAGE_NAME>/<SIGNATURE HASH> - must be registered at https://aka.ms/MobileAppReg" />
```

Whereas our test apps configure this the other way around....
```xml
<data
    android:scheme="msal<client_id>"
    android:host="auth" />
```

Are we expecting that developers will register and configure their broker `redirect_uri` in the app config/manifest? Or that they can use any arbitrary `redirect_uri` and we will compute the value _only_ when connecting to the broker?

**Comment by iambmelt** [comment]
As for resolving the build-scaryness, I suggest we use placeholders that don't involve illegal characters _or_ ship a working sample config that developers can use out-of-the-box**

>** As discussed offline, there are drawbacks to this, such as developers actually _shipping_ these values

**Comment by iambmelt** [comment]
For the default config, I think we should avoid portraying `common` as `tenant_id` - this property is optional and I think in this cases confuses more than aids

```xml
{
  "client_id" : "Register your app at https://aka.ms/MobileAppReg",
  "authorization_user_agent" : "DEFAULT",
  "redirect_uri" : "Register your app at https://aka.ms/MobileAppReg",
  "authorities" : [
    {
      "type": "AAD",
      "audience": {
        "type": "AzureADandPersonalMicrosoftAccount",
        "tenant_id": "common" <-- Not a real tenant!
      }
    }
  ]
}
```

**Comment by iambmelt** [comment]
For the project `build.gradle` -- we should reorder the repositories. Recently Google/AOSP have move dependencies out of `jcenter()` and into `google()`, causing lots of `404s` when `jcenter()` is declared first.

Before:
```groovy
allprojects {
    repositories {
        jcenter()
        mavenCentral()
        maven {
            url 'https://maven.google.com/'
            name 'Google'
        }
    }
}
```

After:
```groovy
allprojects {
    repositories {
        google()
        jcenter()
        mavenCentral()
    }
}
```

**Comment by iambmelt** [comment]
The `buildToolsVersion` shipping in this sample is pretty old

Ships: `com.android.tools.build:gradle:3.1.4`
Latest: `com.android.tools.build:gradle:3.4.0`

**Comment by iambmelt** [comment]
This conditional statement is always `true` based on the assignment.

https://github.com/Azure-Samples/ms-identity-android-native/blob/d91c72fd6e2d3779d5d78702a3cba3550b9c51a0/app/src/main/java/com/azuresamples/msalandroidapp/MainActivity.java#L58-L59

**Comment by iambmelt** [comment]
In the latest tooling versions, these `findViewById` calls no longer need to be casted, the type can be inferred

https://github.com/Azure-Samples/ms-identity-android-native/blob/d91c72fd6e2d3779d5d78702a3cba3550b9c51a0/app/src/main/java/com/azuresamples/msalandroidapp/MainActivity.java#L42-L43

**Comment by iambmelt** [comment]
This array declaration, while valid, is C-style (not Java-style) and generates warnings

https://github.com/Azure-Samples/ms-identity-android-native/blob/d91c72fd6e2d3779d5d78702a3cba3550b9c51a0/app/src/main/java/com/azuresamples/msalandroidapp/MainActivity.java#L25

In Java-style
```java
final static String[] SCOPES = {"https://graph.microsoft.com/User.Read"};
```

**Comment by iambmelt** [comment]
This condition is impossible

https://github.com/Azure-Samples/ms-identity-android-native/blob/d91c72fd6e2d3779d5d78702a3cba3550b9c51a0/app/src/main/java/com/azuresamples/msalandroidapp/MainActivity.java#L73

**Comment by iambmelt** [comment]
Could use the Graph SDK here - might be nice nod to some partners ;)

https://github.com/Azure-Samples/ms-identity-android-native/blob/d91c72fd6e2d3779d5d78702a3cba3550b9c51a0/app/src/main/java/com/azuresamples/msalandroidapp/MainActivity.java#L131-L175

**Comment by NLLAPPS** [comment]
Is auth_config.json required? I got my app working without it just fine.

**Comment by iambmelt** [comment]
We can omit this `throws` declaration, as it cannot happen

https://github.com/Azure-Samples/ms-identity-android-native/blob/d91c72fd6e2d3779d5d78702a3cba3550b9c51a0/app/src/main/java/com/azuresamples/msalandroidapp/MainActivity.java#L161

**Comment by iambmelt** [comment]
'Canceled' or 'Cancelled'? ;-)

https://github.com/Azure-Samples/ms-identity-android-native/blob/d91c72fd6e2d3779d5d78702a3cba3550b9c51a0/app/src/main/java/com/azuresamples/msalandroidapp/MainActivity.java#L263-L264

**Comment by iambmelt** [comment]
This block....

https://github.com/Azure-Samples/ms-identity-android-native/blob/d91c72fd6e2d3779d5d78702a3cba3550b9c51a0/app/src/main/java/com/azuresamples/msalandroidapp/MainActivity.java#L104-L128

I see you've javadoc'd to keep expectations measured - I'm a little wary of the 'sign-out' language

**Comment by danieldobalian** [comment]
@iambmelt @kreedula thanks for the review. Incoming changes with all the fixes. Brian's first 3 comments have no been applied and need to be discussed offline

**Comment by danieldobalian** [comment]
@NLLAPPS the configuration is optional today, but we may make it the only option in the future. Do you have any reason you prefer not to use it? The design here is still open to influence and I'd like to hear your feedback. 

As for the release + portal link to point to the awesome new experience, I'd expect something in the next 0-2 weeks!

**Comment by NLLAPPS** [comment]
This SDK is part of CloudResources module I use in my few of apps. It was a lot easier to provide config in Java with my custom config provider and have main app(s) provide it to the Cloud module .

**Comment by NLLAPPS** [comment]
I would also like to comment on  https://aka.ms/MobileAppReg url.
It loads https://aad.portal.azure.com which I don't have permission to register new application with my outlook account which I currently use to register my clients. Old Live SDK link apps.dev.microsoft.com allows me to register clients with my outlook account.

I can however use https://portal.azure.com/ with my outlook account. 

I am able to use https://aad.portal.azure.com  with my Office 365/Exchange account but then client would be tied to my subscription and would be deleted if I cancel my Office365/Exchange subscription

**Comment by danieldobalian** [comment]
@NLLAPPS thanks for all the info. Can you elaborate on your Cloud Resource module/

The link is not functional yet, but will be once our updated portal ships. For now, it's just a placeholder. For the issue about registering apps on the Azure portal, I would like to understand more. Can you reach out via email so we can chat offline? My email is in my Github profile. In the meantime, the Azure portal will also allow you to register/manage apps with your Outlook account and does NOT require a subscription for Azure AD app registrations.

**Comment by NLLAPPS** [comment]
Re Cloud Resources:
I have couple of apps that supports few cloud services such as OneDrive, Google Drive etc. I maintain one singe cloud upload module that abstracts different cloud services. It makes applying changes to different SDKs easy. 

Re Registration:
I'll email you from m at nllapps.com with screenshots to clarify the issue further

File: README.md [hunk 1]
```diff
@@ -18,15 +18,16 @@ The MSAL Android preview gives your app the ability to begin using the
...
 * Get a token for the Microsoft Graph
-* Refresh a token
+* Device-wide SSO and Conditional Access suport through the Auth Broker
+* Refresh tokens automatically
 * Call the Microsoft Graph
 * Sign out the user
 
 ## Scenario
 
 This app is a multi-tenant app meaning it can be used by any Azure AD tenant or Microsoft Account.  It demonstrates how a developer can build apps to connect with enterprise users and access their Azure + O365 data via the Microsoft Graph.  During the auth flow, end users will be required to sign in and consent to the permissions of the application, and in some cases may require an admin to consent to the app.  The majority of the logic in this sample shows how to auth an end user and make a basic call to the Microsoft Graph.
```
Comment (shoatman) [inline README.md]: Permissions requested by the application

File: README.md [hunk 2]
```diff
@@ -45,43 +46,39 @@ pApp.acquireToken(getActivity(), SCOPES, getAuthInteractiveCallback());
...
+3. Configure the sample code in `auth_config.JSON` and `AndroidManifest.xml`.
+    - Copy and paste the ***MSAL Configuration*** JSON from the Azure portal into `auth_config.JSON`. 
+    - Inside the `AndroidManifest.xml`, replace `android:host` with the same info registered above.
+        - `auth_config.JSON` contains this information as a reference inside the `redirect_uri` field.
 
-3. Select *Build* > *Clean Project*. 
+5. Select *Build* > *Clean Project*. 
 
-4. Select *Run* > *Run 'app'*. Make sure the emulator you're using has Chrome, if it doesn't follow 
+6. Select *Run* > *Run 'app'*. Make sure the emulator you're using has Chrome, if it doesn't follow
```
Comment (shoatman) [inline README.md]: Chrome is no longer a requirement is it?

File: app/src/main/java/com/azuresamples/msalandroidapp/MainActivity.java [hunk 3]
```diff
@@ -22,7 +22,7 @@
 public class MainActivity extends AppCompatActivity {
 
     /* Azure AD v2 Configs */
-    final static String SCOPES [] = {"https://graph.microsoft.com/User.Read"};
+    final static String[] SCOPES = {"https://graph.microsoft.com/User.Read"};
```
Comment (shoatman) [inline app/src/main/java/com/azuresamples/msalandroidapp/MainActivity.java]: MS Graph scopes do not need to be prefixed

File: gradle/wrapper/gradle-wrapper.properties [hunk 4]
```diff
@@ -1,6 +1,6 @@
-#Tue Aug 07 18:42:11 PDT 2018
+#Sun Apr 21 11:48:57 PDT 2019
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.4-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-5.1.1-all.zip
```
Comment (shoatman) [inline gradle/wrapper/gradle-wrapper.properties]: May make sense to share this in the readme for teh sample: https://developer.android.com/studio/releases/gradle-plugin

File: app/src/main/AndroidManifest.xml [hunk 5]
```diff
@@ -30,7 +30,8 @@
 
                 <!--Add in your scheme/host from registered redirect URI-->
                 <data android:scheme="msauth"
-                    android:host="<PACKAGE_NAME>/<SIGNATURE HASH> - must be registered at https://aka.ms/MobileAppReg" />
+                    android:host= "<PACKAGE_NAME> - must be registered at https://aka.ms/MobileAppReg"
+                    android:path= "/<ENCODED_SIGNATURE_HASH - must be registered at https://aka.ms/MobileAppReg>" />
```
Comment (heidijinxujia) [inline app/src/main/AndroidManifest.xml]: Decoded signature hash, not encoded.

File: app/src/main/AndroidManifest.xml [hunk 6]
```diff
@@ -30,7 +30,8 @@
 
                 <!--Add in your scheme/host from registered redirect URI-->
                 <data android:scheme="msauth"
-                    android:host="<PACKAGE_NAME>/<SIGNATURE HASH> - must be registered at https://aka.ms/MobileAppReg" />
+                    android:host= "<PACKAGE_NAME> - must be registered at https://aka.ms/MobileAppReg"
```
Comment (heidijinxujia) [inline app/src/main/AndroidManifest.xml]: The package name must be in lower case.

File: app/src/main/java/com/azuresamples/msalandroidapp/MainActivity.java (hunk 3 above)
Comment (danieldobalian) [reply app/src/main/java/com/azuresamples/msalandroidapp/MainActivity.java]: @iambmelt gave the opposite feedback - this initialization is more C than Java style he taught me. I usually do what you have above as well :).

File: app/src/main/AndroidManifest.xml (hunk 6 above)
Comment (danieldobalian) [reply app/src/main/AndroidManifest.xml]: I think we'll try to apply this when configuring in the portal. Need to discuss on Monday w/ Portal team.

File: app/src/main/AndroidManifest.xml (hunk 5 above)
Comment (danieldobalian) [reply app/src/main/AndroidManifest.xml]: Thank you!

**Pull request merged**